fem/fe/fe_nurbs.cpp

```diff
@@ -398,8 +398,625 @@ void NURBS3DFiniteElement::CalcHessian (const IntegrationPoint &ip,
+                                             Vector &divshape) const
+{
+   kv[0]->CalcShape ( shape_x, ijk[0], ip.x);
+   kv[1]->CalcShape ( shape_y, ijk[1], ip.y);
+
+   kv1[0]->CalcDShape(dshape1_x, ijk[0], ip.x);
+   kv1[1]->CalcDShape(dshape1_y, ijk[1], ip.y);
+
+   int o = 0;
+   for (int j = 0; j <= orders[1]; j++)
+   {
+      const double sy = shape_y(j);
+      for (int i = 0; i <= orders[0]+1; i++, o++)
+      {
+         divshape(o) = dshape1_x(i)*sy;
+      }
+   }
+
+   for (int j = 0; j <= orders[1]+1; j++)
+   {
+      const double dsy1 = dshape1_y(j);
+      for (int i = 0; i <= orders[0]; i++, o++)
+      {
+         divshape(o) = shape_x(i)*dsy1;
+      }
+   }
+}
+
+NURBS_HDiv2DFiniteElement::~NURBS_HDiv2DFiniteElement()
+{
+   if (kv1[0]) { delete kv1[0]; }
+   if (kv1[1]) { delete kv1[1]; }
+}
+
+
+void NURBS_HDiv3DFiniteElement::SetOrder() const
+{
+   orders[0] = kv[0]->GetOrder();
+   orders[1] = kv[1]->GetOrder();
+   orders[2] = kv[2]->GetOrder();
+
+   if (kv1[0]) { delete kv1[0]; }
+   if (kv1[1]) { delete kv1[1]; }
+   if (kv1[2]) { delete kv1[2]; }
+
+   kv1[0] = kv[0]->DegreeElevate(1);
+   kv1[1] = kv[1]->DegreeElevate(1);
+   kv1[2] = kv[2]->DegreeElevate(1);
+
+   shape_x.SetSize(orders[0]+1);
+   shape_y.SetSize(orders[1]+1);
+   shape_z.SetSize(orders[2]+1);
+
+   dshape_x.SetSize(orders[0]+1);
+   dshape_y.SetSize(orders[1]+1);
+   dshape_z.SetSize(orders[2]+1);
+
+   d2shape_x.SetSize(orders[0]+1);
+   d2shape_y.SetSize(orders[1]+1);
+   d2shape_z.SetSize(orders[2]+1);
+
+   shape1_x.SetSize(orders[0]+2);
+   shape1_y.SetSize(orders[1]+2);
+   shape1_z.SetSize(orders[2]+2);
+
+   dshape1_x.SetSize(orders[0]+2);
+   dshape1_y.SetSize(orders[1]+2);
+   dshape1_z.SetSize(orders[2]+2);
+
+   d2shape1_x.SetSize(orders[0]+2);
+   d2shape1_y.SetSize(orders[1]+2);
+   d2shape1_z.SetSize(orders[2]+2);
+
+   order = max(orders[0]+1, max( orders[1]+1, orders[2]+1));
+   dof = (orders[0] + 2)*(orders[1] + 1)*(orders[2] + 1) +
+         (orders[0] + 1)*(orders[1] + 2)*(orders[2] + 1) +
+         (orders[0] + 1)*(orders[1] + 1)*(orders[2] + 2);
+   u.SetSize(dof);
+   du.SetSize(dof);
+   weights.SetSize(dof);
+}
+
+void NURBS_HDiv3DFiniteElement::CalcVShape(const IntegrationPoint &ip,
+                                           DenseMatrix &shape) const
+{
+   kv[0]->CalcShape(shape_x, ijk[0], ip.x);
+   kv[1]->CalcShape(shape_y, ijk[1], ip.y);
+   kv[2]->CalcShape(shape_z, ijk[2], ip.z);
+
+   kv1[0]->CalcShape(shape1_x, ijk[0], ip.x);
+   kv1[1]->CalcShape(shape1_y, ijk[1], ip.y);
+   kv1[2]->CalcShape(shape1_z, ijk[2], ip.z);
+
+   int o = 0;
+   for (int k = 0; k <= orders[2]; k++)
+   {
+      const double sz = shape_z(k);
+      for (int j = 0; j <= orders[1]; j++)
+      {
+         const double sy_sz = shape_y(j)*sz;
+         for (int i = 0; i <= orders[0]+1; i++, o++)
+         {
+            shape(o,0) = shape1_x(i)*sy_sz;
+            shape(o,1) = shape(o,2) = 0.0;
```

Comment:
Setting `shape = 0.0` before the loop could eliminate a few lines of setting individual entries to 0.